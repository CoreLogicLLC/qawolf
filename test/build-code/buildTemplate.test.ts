import {
  buildImports,
  buildTemplate,
  buildValidVariableName,
} from '../../src/build-code/buildTemplate';

describe('buildImports', () => {
  it('includes device when specified', () => {
    expect(buildImports({ device: 'iPhone 11' })).toMatchSnapshot();
  });

  it('imports when typescript', () => {
    expect(buildImports({ useTypeScript: true })).toMatchSnapshot();

    expect(
      buildImports({ device: 'iPhone 11', useTypeScript: true }),
    ).toMatchSnapshot();
  });

  it('requires when not typescript', () => {
    expect(buildImports({})).toMatchSnapshot();
  });

  it('throws an error if device does not exist', () => {
    const testFn = (): string => buildImports({ device: 'unknown' });
    expect(testFn).toThrowError('Device unknown not available in Playwright');
  });
});

describe('buildTemplate', () => {
  const options = {
    name: 'myName',
    url: 'www.qawolf.com',
  };

  it('builds test template', () => {
    let template = buildTemplate(options);
    expect(template).toMatchSnapshot();

    template = buildTemplate({
      ...options,
      device: 'iPhone 11',
    });
    expect(template).toMatchSnapshot();

    template = buildTemplate({
      ...options,
      statePath: 'admin.json',
    });
    expect(template).toMatchSnapshot();

    template = buildTemplate({
      ...options,
      useTypeScript: true,
    });
    expect(template).toMatchSnapshot();

    template = buildTemplate({
      ...options,
      device: 'iPhone 7',
      useTypeScript: true,
    });
    expect(template).toMatchSnapshot();
  });
});

describe('buildValidVariableName', () => {
  it('returns valid variable names as is', () => {
    expect(buildValidVariableName('myScript')).toBe('myScript');
    expect(buildValidVariableName('my_script')).toBe('my_script');
    expect(buildValidVariableName('myScript1')).toBe('myScript1');
    expect(buildValidVariableName('$myScript')).toBe('$myScript');
  });

  it('returns corrected invalid variable names', () => {
    expect(buildValidVariableName('my-script')).toBe('myScript');
    expect(buildValidVariableName('my script')).toBe('myScript');
  });

  it('throws error if variable name cannot be corrected', () => {
    expect(() => buildValidVariableName('return')).toThrowError('return');
    expect(() => buildValidVariableName('1myScript')).toThrowError('1myScript');
  });
});
