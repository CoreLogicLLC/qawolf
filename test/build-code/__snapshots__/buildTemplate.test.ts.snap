// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildImports imports when typescript 1`] = `
"import { Browser, BrowserContext } from \\"playwright\\";
import qawolf from \\"qawolf\\";"
`;

exports[`buildImports imports when typescript 2`] = `
"import { Browser, BrowserContext, devices } from \\"playwright\\";
import qawolf from \\"qawolf\\";
const device = devices[\\"iPhone 11\\"];"
`;

exports[`buildImports includes device when specified 1`] = `
"const { devices } = require(\\"playwright\\");
const qawolf = require(\\"qawolf\\");
const device = devices[\\"iPhone 11\\"];"
`;

exports[`buildImports requires when not typescript 1`] = `"const qawolf = require(\\"qawolf\\");"`;

exports[`buildTemplate builds test template 1`] = `
"const qawolf = require(\\"qawolf\\");

let browser;
let context;

beforeAll(async () => {
  browser = await qawolf.launch();
  context = await browser.newContext();
  await qawolf.register(context);
});

afterAll(async () => {
  await qawolf.stopVideos();
  await browser.close();
});

test(\\"myName\\", async () => {
  await qawolf.create(\\"www.qawolf.com\\");
});"
`;

exports[`buildTemplate builds test template 2`] = `
"const { devices } = require(\\"playwright\\");
const qawolf = require(\\"qawolf\\");
const device = devices[\\"iPhone 11\\"];

let browser;
let context;

beforeAll(async () => {
  browser = await qawolf.launch();
  context = await browser.newContext({ ...device });
  await qawolf.register(context);
});

afterAll(async () => {
  await qawolf.stopVideos();
  await browser.close();
});

test(\\"myName\\", async () => {
  await qawolf.create(\\"www.qawolf.com\\");
});"
`;

exports[`buildTemplate builds test template 3`] = `
"const qawolf = require(\\"qawolf\\");

let browser;
let context;

beforeAll(async () => {
  browser = await qawolf.launch();
  context = await browser.newContext();
  await qawolf.register(context);
});

afterAll(async () => {
  await qawolf.stopVideos();
  await browser.close();
});

test(\\"myName\\", async () => {
  await qawolf.setState(page, \\"admin.json\\");
  await qawolf.create(\\"www.qawolf.com\\");
});"
`;

exports[`buildTemplate builds test template 4`] = `
"import { Browser, BrowserContext } from \\"playwright\\";
import qawolf from \\"qawolf\\";

let browser: Browser;
let context: BrowserContext;

beforeAll(async () => {
  browser = await qawolf.launch();
  context = await browser.newContext();
  await qawolf.register(context);
});

afterAll(async () => {
  await qawolf.stopVideos();
  await browser.close();
});

test(\\"myName\\", async () => {
  await qawolf.create(\\"www.qawolf.com\\");
});"
`;

exports[`buildTemplate builds test template 5`] = `
"import { Browser, BrowserContext, devices } from \\"playwright\\";
import qawolf from \\"qawolf\\";
const device = devices[\\"iPhone 7\\"];

let browser: Browser;
let context: BrowserContext;

beforeAll(async () => {
  browser = await qawolf.launch();
  context = await browser.newContext({ ...device });
  await qawolf.register(context);
});

afterAll(async () => {
  await qawolf.stopVideos();
  await browser.close();
});

test(\\"myName\\", async () => {
  await qawolf.create(\\"www.qawolf.com\\");
});"
`;
